### PR TITLE
Refactor visibility_detector to avoid forcing compositing.

### PR DESCRIPTION
**Pull request by dnfield**
Instead, use methods on RenderObject and new framework API to
listen to compositing and determine the transform/clip.

Removes an O(N^2) algorithm from the ancestor walk.

Deletes the custom layer that is no longer used.

Minor update to main.dart to be more desktop friendly.

Tests adjusted to use new API. Removed API has no usages
in internal codesearch.

Depends on https://github.com/flutter/flutter/pull/103378

### CODE REVIEW

File: packages/visibility_detector/example/lib/main.dart
Comment (xster) [inline packages/visibility_detector/example/lib/main.dart:77]: is this related?

Comment (dnfield) [reply packages/visibility_detector/example/lib/main.dart:77]: This makes the framework stop complaining when you run the example app in desktop mode. It's not essential for this change.

Comment (dnfield) [reply packages/visibility_detector/example/lib/main.dart:77]: (No, it's not related)

**Comment by dnfield** [comment]
Ok. In some local benchmarking on a Pixel 4 this is showing some improvements over the old version now. I've made the following changes from initial:

- Defer actually calculating visibility (original version of this patch was deferring updates about visibility but eagerly calculating)
- Make sure to unregister callbacks when paint is called and add tests (original patch was leaving old callbacks around which made compositing slow down).
- Cache ancestor lookups. Testing in the demo app and customer: money's simplified app show a good number of hits on this cache.

The caching logic can probably be further improved. The lookups would benefit from using the layer tree, but that's currently blocked on https://github.com/flutter/flutter/issues/103469

**Comment by dnfield** [comment]
I'll mark this as ready for review once the framework PR has landed and I can put a semi-meaningful framework version into the SDK constraint.

**Comment by dnfield** [comment]
Moto g4 at head:

```
  "average_frame_build_time_millis": 2.9323076923076914,
  "90th_percentile_frame_build_time_millis": 4.431,
  "99th_percentile_frame_build_time_millis": 5.282,
  "worst_frame_build_time_millis": 6.973,
  "missed_frame_build_budget_count": 0,
  "average_frame_rasterizer_time_millis": 10.860470588235296,
  "90th_percentile_frame_rasterizer_time_millis": 15.983,
  "99th_percentile_frame_rasterizer_time_millis": 45.917,
  "worst_frame_rasterizer_time_millis": 45.917,
  "missed_frame_rasterizer_budget_count": 5,
  "frame_count": 52,
  "frame_rasterizer_count": 51,
  "new_gen_gc_count": 2,
  "old_gen_gc_count": 0,
  "average_vsync_transitions_missed": 1.4166666666666667,
  "90th_percentile_vsync_transitions_missed": 2.0,
  "99th_percentile_vsync_transitions_missed": 3.0,
  "average_vsync_frame_lag": 563.8333333333334,
  "90th_percentile_vsync_frame_lag": 1266.0,
  "99th_percentile_vsync_frame_lag": 1659.0,
  "average_layer_cache_count": 1.6538461538461537,
  "90th_percentile_layer_cache_count": 3.0,
  "99th_percentile_layer_cache_count": 4.0,
  "worst_layer_cache_count": 4.0,
  "average_layer_cache_memory": 0.15911871153846155,
  "90th_percentile_layer_cache_memory": 0.441193,
  "99th_percentile_layer_cache_memory": 0.512878,
  "worst_layer_cache_memory": 0.512878,
  "average_picture_cache_count": 10.903846153846153,
  "90th_percentile_picture_cache_count": 15.0,
  "99th_percentile_picture_cache_count": 16.0,
  "worst_picture_cache_count": 16.0,
  "average_picture_cache_memory": 7.179940769230776,
  "90th_percentile_picture_cache_memory": 9.070084,
  "99th_percentile_picture_cache_memory": 9.140945,
  "worst_picture_cache_memory": 9.140945,
  "total_ui_gc_time": 4.757,

```

Moto G4 with this patch:

```
  "average_frame_build_time_millis": 3.1556440677966098,
  "90th_percentile_frame_build_time_millis": 4.479,
  "99th_percentile_frame_build_time_millis": 5.321,
  "worst_frame_build_time_millis": 5.573,
  "missed_frame_build_budget_count": 0,
  "average_frame_rasterizer_time_millis": 9.560338983050848,
  "90th_percentile_frame_rasterizer_time_millis": 13.052,
  "99th_percentile_frame_rasterizer_time_millis": 34.242,
  "worst_frame_rasterizer_time_millis": 38.225,
  "missed_frame_rasterizer_budget_count": 5,
  "frame_count": 59,
  "frame_rasterizer_count": 59,
  "new_gen_gc_count": 2,
  "old_gen_gc_count": 0,
"average_vsync_transitions_missed": 1.3076923076923077,
  "90th_percentile_vsync_transitions_missed": 2.0,
  "99th_percentile_vsync_transitions_missed": 2.0,
  "average_vsync_frame_lag": 554.8,
  "90th_percentile_vsync_frame_lag": 1339.0,
  "99th_percentile_vsync_frame_lag": 1833.0,
  "average_layer_cache_count": 1.6,
  "90th_percentile_layer_cache_count": 3.0,
  "99th_percentile_layer_cache_count": 4.0,
  "worst_layer_cache_count": 4.0,
  "average_layer_cache_memory": 0.1566712166666669,
  "90th_percentile_layer_cache_memory": 0.441193,
  "99th_percentile_layer_cache_memory": 0.512878,
  "worst_layer_cache_memory": 0.512878,
  "average_picture_cache_count": 9.6,
  "90th_percentile_picture_cache_count": 14.0,
  "99th_percentile_picture_cache_count": 15.0,
  "worst_picture_cache_count": 15.0,
  "average_picture_cache_memory": 5.347194483333333,
  "90th_percentile_picture_cache_memory": 6.723358,
  "99th_percentile_picture_cache_memory": 7.494987,
  "worst_picture_cache_memory": 7.494987,
  "total_ui_gc_time": 4.762,

**Comment by jonahwilliams** [comment]
those are great numbers

**Comment by dnfield** [comment]
Looking into some internal test failures.

**Comment by dnfield** [comment]
https://github.com/flutter/flutter/pull/103768/ and https://github.com/flutter/flutter/pull/103748 are blocking this right now

**Comment by dnfield** [comment]
Latest change is down to ~20 internal failures (not including minor scuba changes) as long as the linked flutter/flutter patches are patched in. Will keep looking into that.

**Comment by dnfield** [comment]
I'm down to one more issue on this, where it seemslike sometimes I'm losing track of a scroll offset or transform in the tree.

**Comment by dnfield** [comment]
This results in a number of small golden changes in g3, mainly because of removing layers/pixel snapping.

There are a couple cases where images change: one is a semantics debugger where the colors are changing, and another looks like the widget acutally loads more - but all the visibility detection related tests are passing.

**Comment by dnfield** [comment]
I don' thave permission to add reviewers in this repo or to land. I'd appreciate any feedback from @jonahwilliams and @goderbauer on this.

**Comment by dnfield** [comment]
The internal CL and current golden changes are at cl/450011940 - you can see the associated tap run there if you are a Googler.

File: packages/visibility_detector/lib/src/render_visibility_detector.dart
Comment (goderbauer) [inline packages/visibility_detector/lib/src/render_visibility_detector.dart:86]: The VisibilityInfo object contains the newly added `screenRect` value. It appears that there is a code pass where the `screenRect` would change, but the `onVisibilityChanged` is not called (because of the early return above when matchesVisibility is true) to inform the owner about the callback about it, leaving them with potential outdated information. This could be surprising. 

Not sure what the motivation was to add screenRect to that data class, but having an outdated value there could be a source of future bugs.

File: packages/visibility_detector/lib/src/render_visibility_detector.dart
Comment (goderbauer) [inline packages/visibility_detector/lib/src/render_visibility_detector.dart:221]: Do we need to "forget" this RO in dispose so the visibility logic is no longer triggered for it?

**Review by goderbauer — COMMENTED** [review]
Overall, this seems like a good approach.

File: packages/visibility_detector/lib/src/render_visibility_detector.dart
Comment (dnfield) [reply packages/visibility_detector/lib/src/render_visibility_detector.dart:86]: Ahhh, I wanted to get rid of a global static dictionary. 

The only thing that uses these are unit tests in this library. I suppose I could just drop that entirely. Otherwise we'd have to emit changes when just the screen rect changes, which maybe is fine but maybe is a small change in behavior that will cause problems for google consumers.

File: packages/visibility_detector/lib/src/render_visibility_detector.dart
Comment (dnfield) [reply packages/visibility_detector/lib/src/render_visibility_detector.dart:130]: Done

File: packages/visibility_detector/lib/src/render_visibility_detector.dart
Comment (dnfield) [reply packages/visibility_detector/lib/src/render_visibility_detector.dart:152]: I'm not sure how we can do that - we might, for example, be in the middle of a frame and that still means this callback will fire, but `hasScheduledFrame` will be false.

This comment is legacy from the layer code. I suppose I could delete it?

File: packages/visibility_detector/lib/src/render_visibility_detector.dart
Comment (dnfield) [reply packages/visibility_detector/lib/src/render_visibility_detector.dart:160]: This is unchanged from the original. 

I think the idea here is that if it were to use a `RestartableTimer` it'd be stuck with the originally supplied duration, as opposed to creating a new timer each time with the right duration.

File: packages/visibility_detector/lib/src/render_visibility_detector.dart
Comment (dnfield) [reply packages/visibility_detector/lib/src/render_visibility_detector.dart:221]: Yes, that may be our only chance to let clients know that the RO is no longer visible (assuming it ever was). There are tests that fail if you add a forget here.

File: packages/visibility_detector/lib/src/render_visibility_detector.dart
Comment (dnfield) [reply packages/visibility_detector/lib/src/render_visibility_detector.dart:246]: Good idea.

File: packages/visibility_detector/lib/src/render_visibility_detector.dart
Comment (dnfield) [reply packages/visibility_detector/lib/src/render_visibility_detector.dart:253]: It's what the original code used.

In the framework, `semanticBounds` and `paintBounds` are always the same, and for box classes they're `Offset.zero & size` :)

File: packages/visibility_detector/lib/src/render_visibility_detector.dart
Comment (dnfield) [reply packages/visibility_detector/lib/src/render_visibility_detector.dart:119]: Done - removed the parameter from this entirely and just have an abstract `Rect get bounds` on the mixin.

File: packages/visibility_detector/lib/src/render_visibility_detector.dart
Comment (dnfield) [reply packages/visibility_detector/lib/src/render_visibility_detector.dart:246]: Done, as mentioned elsehwere moved this to a `Rect get bounds;` on the mixin, which simplified things a bit.

**Comment by dnfield** [comment]
@jamesderlin I'm also interested in any feedback you have, if you have any to share :)

File: packages/visibility_detector/lib/src/render_visibility_detector.dart
Comment (goderbauer) [inline packages/visibility_detector/lib/src/render_visibility_detector.dart:112]: Wondering, why does forget not also clear the lastVisibility for the key? Is there a use case where you want to forget, but still keep that historical information?

File: packages/visibility_detector/lib/src/render_visibility_detector.dart
Comment (goderbauer) [inline packages/visibility_detector/lib/src/render_visibility_detector.dart:119]: nit: maybe make the parameters a required optional one so this call side reads a little better...

File: packages/visibility_detector/lib/src/render_visibility_detector.dart
Comment (goderbauer) [inline packages/visibility_detector/lib/src/render_visibility_detector.dart:128]: nit: change 0 to null?

File: packages/visibility_detector/lib/src/render_visibility_detector.dart
Comment (dnfield) [reply packages/visibility_detector/lib/src/render_visibility_detector.dart:86]: So a case where I think you're right: if the widget resizes, but the resizing happens outside of the current clip. The visible bounds won't change but the widget bounds would.

I'm going to remove this as it's only used in tests.

File: packages/visibility_detector/lib/src/render_visibility_detector.dart
Comment (goderbauer) [inline packages/visibility_detector/lib/src/render_visibility_detector.dart:280]: similar to above:

```suggestion
  }) : assert(key != null),
        _onVisibilityChanged = onVisibilityChanged,
        super(child);
```

File: packages/visibility_detector/lib/src/render_visibility_detector.dart
Comment (goderbauer) [reply packages/visibility_detector/lib/src/render_visibility_detector.dart:86]: This remains my main concern as I can already see how people will rely on this and then get surprised if it doesn't update properly. I'd either remove this feature alltogether, or - if it is only used in test - make it a debug only feature that throws in release when accessed.

**Review by goderbauer — COMMENTED** [review]
The `screenRect` property is still of concern to me. Everything else is just nits..

File: packages/visibility_detector/lib/src/render_visibility_detector.dart
Comment (dnfield) [reply packages/visibility_detector/lib/src/render_visibility_detector.dart:280]: Doesn't work - because the private field is on a mixin. Maybe this is a dart bug?

File: packages/visibility_detector/lib/src/render_visibility_detector.dart
Comment (dnfield) [reply packages/visibility_detector/lib/src/render_visibility_detector.dart:128]: Done

**Comment by goderbauer** [comment]
> Unfortunately, CI is not set up for this repo...

That IS unfortunate. Might be time to set that up?

File: packages/visibility_detector/lib/src/render_visibility_detector.dart
Comment (dnfield) [reply packages/visibility_detector/lib/src/render_visibility_detector.dart:119]: Going to make it an optional optional :)

File: packages/visibility_detector/lib/src/render_visibility_detector.dart
Comment (goderbauer) [reply packages/visibility_detector/lib/src/render_visibility_detector.dart:280]: I see. Interesting.
Maybe add the assert, though just for symmetry with the other constructor above..

**Review by goderbauer — APPROVED** [review]
With `screenRect` removed, this LGTM.

**Comment by dnfield** [comment]
Ahh actions aren't enabled. I'll revert my commit.

**Pull request merged**
